lib/src/position.dart

```diff
@@ -2179,16 +2449,38 @@ class Horde extends Position<Horde> {
   }
 }
 
+enum EndType {
+  decisive,
+  stalemate,
+  impasse,
+  impasseMoreMaterial,
+}
+
 /// The outcome of a [Position]. No `winner` means a draw.
 @immutable
 class Outcome {
-  const Outcome({this.winner});
+  const Outcome({this.winner, this.endType});
 
   final Side? winner;
-
-  static const whiteWins = Outcome(winner: Side.white);
-  static const blackWins = Outcome(winner: Side.black);
-  static const draw = Outcome();
+  final EndType? endType;
+
+  static const whiteWins =
+      Outcome(winner: Side.white, endType: EndType.decisive);
+  static const blackWins =
+      Outcome(winner: Side.black, endType: EndType.decisive);
+  static const stalemate = Outcome(endType: EndType.stalemate);
+  static const draw = Outcome(endType: EndType.impasse);
```

Comment:
should we call it `impasse` instead?

Reply:
As part of the Chess♯ project we wanted to specify the different non-decisive outcomes and (following Lasker) award partial victories where they are merited. In tournaments, partial victories is a better system than using tie-breaking formulas (strength of opposition, etc.)... and it makes games more interesting for spectators, too.

We call it a **Stalemate** when the king is not currently being attacked, but all legal moves would put the king in jeopardy. This is different from the technical definition in classical chess where it's your turn and there are no legal moves... but the representation on the board is essentially the same.

We call it an **Impasse** when there are 50 moves without a capture or pawn-push. 

**Draw** is a broad word, but that's the word codified in classical chess. In classical chess, I could stalemate my opponent while having four queens on the board, and that's scored as a "draw" just like a 75-move endgame that slowly grinds down to king vs king. Both are "draws" but they aren't both "impasses". It's also a "draw" if [two grandmasters just don't feel like playing today]( https://www.chessgames.com/perl/chessgame?gid=1067422). (Bobby Fischer famously accused the Russians of colluding against him by using quick draws-by-agreement. They would draw with each other and save their mental energy for their games against him.)

Chess♯ doesn't have draws. No draws by repetition nor draws by agreement. The game plays on until a king is captured, a king is stalemated, or 50 moves are made without a pawn-move or piece-capture. A player may suggest to their opponent that the game is in a state such that an impasse is inevitable in the next 50 moves. If the opponent agrees, the game is ruled an impasse. (Our term for this situation is **fast-fifty**—as in “I propose a fast-fifty” and “I accept/reject the fast-fifty”.) But ruling it as an impasse still means that one side is awarded a partial victory (7–3, 3–7, or 4–6).

[Pages 9 & 10 of the Chess♯ Rules and FAQ has more about draws, stalemates, and impasses.](https://chess-sharp.games/ChessSharp_Rules.pdf)

Over-the-board, we don't even use insufficient material as an end condition, but I decided to automatically rule those games as an impasse here in dartchess because it's way easier to obnoxiously abandon a game online than it is in person.

-----

So while it is definitionally inaccurate to map all classical chess "draws" to EndType.Impasse, the practical purpose of the EndTypes is to output the right game scores to the pgn files. Conflating all chess draws to impasses works in that respect.

If you wanted to track more end types in classical chess you could add EndType.ThreeMoveRepetition, EndType.DrawByAgreement, etc. But unless that info goes into the pgns, I don't see a reason to do that.